File: CSharpCodeBase/entities/mockup/mockupentity.cs
using System;
 
 using System.Collections.Generic;
 
 using System.Linq;
 
 using System.Text;
 
 using UnityEngine;
 
 using System.Collections;
 
 using MainGame.core;
 
 using UnityEngine.EventSystems;
 
 namespace MainGame{
 public class mockupentity :unityexistsentity{
 public void init (self, gameObject);
     UnityExistsEntity.init(self, gameObject);
     this.count = 0;
     this.oldCount = 0;
     this.enable = false;
     this.lerpEnd = true;
     this.luaMapper:SetAlwaysVisible(true);
     this.oldPosition = Input.RaycastMouseOnTerrain();
     Transform.SetMaterialColor(this.gameObject, 0, 0.5, 0, 0.5);
 }})
 public void Enable(){
   this.enable = true;
 }
 public void Disable(){
   this.enable = false;
   self:Destroy();
 }
 public void Update(){
   UnityExistsEntity.Update(self);
   
   if(this.enable == true  ){ 
     this.position = Input.RaycastMouseOnTerrain();
     if(math.abs(this.oldPosition.x - this.position.x) > 0.05  ||  math.abs(this.oldPosition.z - this.position.z) > 0.05  ){ 
       Transform.Lerp(this.gameObject, this.position.x, this.position.y, this.position.z);
       this.lerpEnd = false;
     }else{if this.lerpEnd == false  ){ 
       if(math.abs(this.oldPosition.x - this.gameObject.transform.position.x)<0.1  &&  math.abs(this.gameObject.transform.position.z - this.oldPosition.z)<0.1  ){ 
         this.lerpEnd = true;
       }else{
         Transform.Lerp(this.gameObject, this.oldPosition.x, this.oldPosition.y, this.oldPosition.z) ;
       }
     }
     this.oldPosition = this.position;
   }
   if(this.oldCount != this.count  ){ 
       this.oldCount = this.count;
       if(this.count == 0  ){ 
         Transform.SetMaterialColor(this.gameObject, 0, 0.5, 0, 0.5);
       }else{
         Transform.SetMaterialColor(this.gameObject, 0.5, 0, 0, 0.5);
       }
   }
 }
 public void Rotate(angle){
   RigidbodyUtils.Rotate(this.gameObject, angle);
 }
 public void MoveUpTransform(speed){
   RigidbodyUtils.MoveTransform(this.gameObject, speed, 0);
 }
 public void GetPosition(){
   return Vector3(this.gameObject.transform.position.x, this.gameObject.transform.position.y, this.gameObject.transform.position.z);
 }
 public void MoveRightTransform(speed){
   RigidbodyUtils.MoveTransform(this.gameObject, 0, speed);
 }
 public void OnCollisionEnter(targetEntity){
     this.count = this.count + 1;
 }
 public void OnCollisionExit(targetEntity){
     this.count = this.count - 1;
 }
 }}

File: CSharpCodeBase/entities/mockup/mockupentity.lua
local Class = require 'utils.hump.class'
local UnityExistsEntity = require 'unity.unityentity'
local Vector3 = require 'utils.hump.vector3'

local MockupEntity = Class({__includes=UnityExistsEntity, init = function (self, gameObject)
    UnityExistsEntity.init(self, gameObject)
    self.count = 0
    self.oldCount = 0
    self.enable = false
    self.lerpEnd = true
    self.luaMapper:SetAlwaysVisible(true)
    self.oldPosition = Input.RaycastMouseOnTerrain()
    Transform.SetMaterialColor(self.gameObject, 0, 0.5, 0, 0.5)
end})

function MockupEntity:Enable()
  self.enable = true
end

function MockupEntity:Disable()
  self.enable = false
  self:Destroy()
end

function MockupEntity:Update()
  UnityExistsEntity.Update(self)
  
  if self.enable == true then
    self.position = Input.RaycastMouseOnTerrain()
    if math.abs(self.oldPosition.x - self.position.x) > 0.05 or math.abs(self.oldPosition.z - self.position.z) > 0.05 then
      Transform.Lerp(self.gameObject, self.position.x, self.position.y, self.position.z)
      self.lerpEnd = false
    elseif self.lerpEnd == false then
      if math.abs(self.oldPosition.x - self.gameObject.transform.position.x)<0.1 and math.abs(self.gameObject.transform.position.z - self.oldPosition.z)<0.1 then
        self.lerpEnd = true
      else
        Transform.Lerp(self.gameObject, self.oldPosition.x, self.oldPosition.y, self.oldPosition.z) 
      end
    end
    self.oldPosition = self.position
  end
  if self.oldCount ~= self.count then
      self.oldCount = self.count
      if self.count == 0 then
        Transform.SetMaterialColor(self.gameObject, 0, 0.5, 0, 0.5)
      else
        Transform.SetMaterialColor(self.gameObject, 0.5, 0, 0, 0.5)
      end
  end
end

function MockupEntity:Rotate(angle)
  RigidbodyUtils.Rotate(self.gameObject, angle)
end

function MockupEntity:MoveUpTransform(speed)
  RigidbodyUtils.MoveTransform(self.gameObject, speed, 0)
end

function MockupEntity:GetPosition()
  return Vector3(self.gameObject.transform.position.x, self.gameObject.transform.position.y, self.gameObject.transform.position.z)
end

function MockupEntity:MoveRightTransform(speed)
  RigidbodyUtils.MoveTransform(self.gameObject, 0, speed)
end

function MockupEntity:OnCollisionEnter(targetEntity)
    self.count = self.count + 1
end

function MockupEntity:OnCollisionExit(targetEntity)
    self.count = self.count - 1
end

return MockupEntity